<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>‘Statistical method you should know’: regression calibration</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="2023-02-16-blog-post-2_files/libs/clipboard/clipboard.min.js"></script>
<script src="2023-02-16-blog-post-2_files/libs/quarto-html/quarto.js"></script>
<script src="2023-02-16-blog-post-2_files/libs/quarto-html/popper.min.js"></script>
<script src="2023-02-16-blog-post-2_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="2023-02-16-blog-post-2_files/libs/quarto-html/anchor.min.js"></script>
<link href="2023-02-16-blog-post-2_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="2023-02-16-blog-post-2_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="2023-02-16-blog-post-2_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="2023-02-16-blog-post-2_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="2023-02-16-blog-post-2_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">‘Statistical method you should know’: regression calibration</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<p>Random measurement errors associated with short-term dietary assessment instruments (e.g., 24-hour dietary recalls and food records) may cause <a href="https://didierbrassard.github.io/posts/2022/11/blog-post-6/">unexpected bias</a> depending on study objective and target statistic. Fortunately, there are well established methods to mitigate the impact of random errors. One of these methods is called <strong>regression calibration</strong>. In this blog, I introduce the methods and show how it can be applied to a simple nutrition analysis.</p>
<section id="what-is-regression-calibration" class="level1">
<h1>What is ‘regression calibration’ ?</h1>
<p>Regression calibration is one of the statistical methods available to correct the bias due to random errors. The method traditionally requires one or many repeated measurements (Kirkpatrick et al.&nbsp;2022). The method consists of the regression of a repeat measurement on the first measurement using a linear regression model. This measurement error model allows to obtain a correction factor which can then be used to calibrate the association of interest.</p>
<p>Regression calibration was first used to obtain unbiased regression coefficient estimate in a linear regression of a continuous outcome on a covariate measured with error, but the method also works well for logistic or Cox regression models (Keogh et al.&nbsp;2020).</p>
<p>I want to point out that comprehensive methods have been developed for nutrition analyses, including the <a href="https://didierbrassard.github.io/teaching/2022-12-nci-1">National Cancer Institute (NCI) methods</a>. Example applications for the National Cancer Institute methods are available at <a href="https://prevention.cancer.gov/research-groups/biometry/measurement-error-impact/software-measurement-error">Software for Measurement Error in Nutrition Research</a>. The NCI methods also cover a wide range of application and are not limited to regression models. Informally, the NCI methods are very fancy and powerful applications of regression calibration designed for a variety of analysis.</p>
<section id="requirements-and-assumptions" class="level2">
<h2 class="anchored" data-anchor-id="requirements-and-assumptions">Requirements and assumptions</h2>
<p>The regression calibration method has statistical assumptions and requirements. In the list below, <span class="math inline">\(i\)</span> refers to a given individual and <span class="math inline">\(j\)</span> refers to a given short-term measurement of dietary intakes (e.g., a 24-h dietary recall or food records). Regression calibration:</p>
<ol type="1">
<li>Requires at least 50 repeated measurements over non-consecutive days (<span class="math inline">\(n_{i,j=2}\ge 50\)</span>; Kirkpatrick et al.&nbsp;2022 for details);</li>
<li>Requires satisfying the classical measurement error model assumptions:
<ol type="1">
<li>Random errors are normally distributed <span class="math inline">\(E_{j} \sim N(\mu=0,\sigma_{\epsilon}^{2})\)</span>;</li>
<li>Random errors across repeat measurements are independent (non-consecutive days) <span class="math inline">\(E_{j=1} \perp E_{j=2}\)</span>;</li>
<li>Random errors are independent of covariates included, if any <span class="math inline">\(E_{ij} \perp Z_{i}\)</span>;</li>
</ol></li>
<li>The random errors structure (i.e., the variation of intakes from one day to another) is generalizable to the full sample or specific subgroup (e.g., by sex);</li>
</ol>
<p>Applying a measurement error correction method also requires us to assume that the self-reported dietary intakes are unbiased. In other words, that self-reported dietary intakes measured with the short-term dietary assessment instrument are the true intakes (Kirkpatrick et al.&nbsp;2022). We know this assumption is not exactly correct (people tend to underreport their intakes), but it is a working assumption.</p>
</section>
</section>
<section id="example-application-one-independent-variable-measured-with-error" class="level1">
<h1>Example application: one independent variable measured with error</h1>
<p>A common study objective is to estimate the relationship between a dietary factor, as independent variable, and an outcome, as dependent variable. For a continuous outcome, the estimand is <span class="math inline">\(E(Y|X)\)</span> or the expected value of <span class="math inline">\(Y\)</span> given <span class="math inline">\(X\)</span>. For such analysis, we could be interested in the relationship between intake of vegetables and fruits (servings/day) and low-density lipoproteins (LDL) cholesterol concentrations as outcome; an association was found previously (Djousse et al.&nbsp;2004). For this example, we will estimate the expected LDL cholesterol concentrations based on the intake of vegetables and fruits in a sample of the population (<span class="math inline">\(n=5000\)</span>).</p>
<p>We could measure vegetables and fruits intake with 24-h dietary recall data since this instrument has minimal systematic error (Thompson et al.&nbsp;2015). Knowing that 24-hour dietary recalls are affected by random errors, we collect a second 24-hour dietary recall to apply correction methods. Of note, to isolate the impact of random errors, I assume that LDL cholesterol concentrations are measured perfectly and that there is no source of bias (no confounding, no selection bias, no differential errors,…).</p>
<p>To assess the relationship between vegetables and fruits and LDL cholesterol, we would use a linear regression model corresponding to <span class="math inline">\(Y_i=\beta_0 + \beta_{X_i} X_i +\epsilon_i\)</span>. However, a naive analysis ignoring random errors would rather have the estimand <span class="math inline">\(E(Y|W_j)\)</span> and model equation <span class="math inline">\(Y_i=\beta\prime_0 + \beta\prime_{W_i} W_i +\epsilon_i\)</span> .</p>
<section id="generate-data" class="level2">
<h2 class="anchored" data-anchor-id="generate-data">Generate data</h2>
<p>In this simulation, each increase of 1 serving of vegetables and fruits will decrease LDL cholesterol concentrations by 0.10 mmol/L. <strong>-10</strong> mmol/L is the true regression coefficient <span class="math inline">\(\beta_{X}\)</span> for this association.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#           Prepare simulated measures           #</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># generate the 'true' value (no random errors)</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  X  <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">5000</span>, <span class="at">mean=</span><span class="dv">5</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># generate independent random errors</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  E1 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">5000</span>, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="fl">1.3</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="do">## change seed that have different values of random error</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  E2 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">5000</span>, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="fl">1.3</span>)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Add random errors to the 'true' value to derive X measured "on a given day"</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  W1 <span class="ot">&lt;-</span> X<span class="sc">+</span>E1</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>  W2 <span class="ot">&lt;-</span> X<span class="sc">+</span>E2</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    <span class="do">## For plausibility, truncate negative</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    <span class="do">## values at 0 (i.e., cant have negative intakes)</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    W1 <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(W1<span class="sc">&lt;</span><span class="dv">0</span>,<span class="dv">0</span>,W1)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>    W2 <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(W2<span class="sc">&lt;</span><span class="dv">0</span>,<span class="dv">0</span>,W2)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="co"># generate a linear relationship based on X</span></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>  b0 <span class="ot">&lt;-</span> <span class="fl">3.0</span> <span class="co"># hypothetical average LDL cholesterol in this sample</span></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  b1 <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fl">0.10</span> <span class="co"># hypothetical relationship between vegetables and fruits and cholesterol</span></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>  EQ <span class="ot">&lt;-</span>  b0 <span class="sc">+</span> b1 <span class="sc">*</span> X</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>  Y  <span class="ot">&lt;-</span>  <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">5000</span>, EQ, <span class="at">sd=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The proportion of total variance due to random errors for vegetables and fruits intakes is approximately 62% for <span class="math inline">\(W_1\)</span> and 61% for <span class="math inline">\(W_2\)</span>. This is consistent with dietary intake data on a given day where a majority of the total variance is simply due to within-individual variations from one day to another (i.e., random errors). In other words, the majority of the variance for dietary intakes measured on a given day is caused by people eating a bit differently each day for random reasons.</p>
</section>
<section id="assess-relationship-using-a-linear-regression-model" class="level2">
<h2 class="anchored" data-anchor-id="assess-relationship-using-a-linear-regression-model">Assess relationship using a linear regression model</h2>
<p>First, I assess the relationship based on the naive analysis which considers the measured vegetables and fruits <strong>on a given day</strong> as if it reflects long-term/usual intakes. Then, I repeat the analysis with the <strong>true</strong> value, i.e., vegetables and fruits intake measured without errors that correctly reflect long-term/usual intakes.</p>
<p>Typically, we would not know the true vegetables and fruits intakes. This would require observing the actual intakes of many individuals over may days or months! Plus, had we known the true intakes, we would not need to apply measurement error correction as there would be no errors to correct.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">#        Association for naive analysis          #</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Model for E(Y|W)</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>naive <span class="ot">&lt;-</span> <span class="fu">lm</span>(Y <span class="sc">~</span> W1)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Output model parameters</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(parameters)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  naive_param <span class="ot">&lt;-</span> </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    parameters<span class="sc">::</span><span class="fu">model_parameters</span>(naive)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>naive_param</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Parameter   | Coefficient |       SE |         95% CI | t(4998) |      p
------------------------------------------------------------------------
(Intercept) |        2.69 |     0.02 | [ 2.65,  2.74] |  117.43 | &lt; .001
W1          |       -0.04 | 4.36e-03 | [-0.05, -0.03] |   -9.16 | &lt; .001</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#                True association                #</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Model for E(Y|X)</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>true <span class="ot">&lt;-</span> <span class="fu">lm</span>(Y <span class="sc">~</span> X)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Output model parameters</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>true_param <span class="ot">&lt;-</span> </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  parameters<span class="sc">::</span><span class="fu">model_parameters</span>(true)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>true_param</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Parameter   | Coefficient |       SE |         95% CI | t(4998) |      p
------------------------------------------------------------------------
(Intercept) |        3.05 |     0.04 | [ 2.98,  3.12] |   86.81 | &lt; .001
X           |       -0.11 | 6.89e-03 | [-0.13, -0.10] |  -16.24 | &lt; .001</code></pre>
</div>
</div>
<section id="regression-coefficient-of-betaprime_w_j-vs.-beta_x" class="level3">
<h3 class="anchored" data-anchor-id="regression-coefficient-of-betaprime_w_j-vs.-beta_x">Regression coefficient of <span class="math inline">\(\beta\prime_{W_j}\)</span> vs.&nbsp;<span class="math inline">\(\beta_X\)</span></h3>
<p>The naive analysis indicates that a 1-serving increase in vegetables and fruits is associated with a 0.04 mmol/L lower LDL cholesterol concentrations (95%CI, -0.05, -0.03). However, the analysis based on true intakes rather indicates that a 1-serving increase in vegetables and fruits is associated with a 0.11 mmol/L lower LDL cholesterol concentrations (95%CI, -0.13, -0.10), which is consistent with the simulated association. The naive analysis shows a 2.8-fold attenuation of the true relationship. This observation is consistent with theory regarding the impact of one independent variable measured with (classical) random errors on an estimated regression coefficient (Keogh et al.&nbsp;2020).</p>
</section>
<section id="standard-error-of-betaprime_w_j-vs.-beta_x" class="level3">
<h3 class="anchored" data-anchor-id="standard-error-of-betaprime_w_j-vs.-beta_x">Standard error of <span class="math inline">\(\beta\prime_{W_j}\)</span> vs.&nbsp;<span class="math inline">\(\beta_X\)</span></h3>
<p>A noteworthy observation is that the standard error of the naive analysis is 1.6-fold <strong>smaller</strong> than the standard error of the true simulated association. Consequently, the estimated confidence interval for the naive analysis is narrower than the confidence interval of the true simulated association. The naive association not only underestimates the true association effect size, but it also produces an incorrectly narrow (and falsely precise) confidence interval.</p>
</section>
</section>
<section id="measurement-error-correction-with-regression-calibration" class="level2">
<h2 class="anchored" data-anchor-id="measurement-error-correction-with-regression-calibration">Measurement error correction with regression calibration</h2>
<section id="descriptive-statistics-and-overview" class="level3">
<h3 class="anchored" data-anchor-id="descriptive-statistics-and-overview">Descriptive statistics and overview</h3>
<p>It is good practice to look at the raw data before further analyses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">#      Combine measured values with outcome      #</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create data with only measured vegetables and fruits (VF) and outcome</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>VF_simstudy <span class="ot">&lt;-</span> </span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">VF_day1 =</span> W1,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">VF_day2 =</span> W2,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">LDL_cholesterol =</span> Y)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co"># show 5 first rows</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(VF_simstudy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   VF_day1  VF_day2 LDL_cholesterol
1 2.402261 3.207557        1.697239
2 6.001527 5.423947        2.232436
3 1.982719 6.228570        2.406524
4 8.128996 5.125792        2.507730
5 6.782459 5.225180        2.216356
6 2.570475 4.351678        2.046728</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co">#             Descriptive statistics             #</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Intake on first assessment (on a given day)</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="do">## mean intakes</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(VF_simstudy<span class="sc">$</span>VF_day1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.985152</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## range (min to max)</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(VF_simstudy<span class="sc">$</span>VF_day1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  0.00000 10.93836</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## histogram</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(VF_simstudy<span class="sc">$</span>VF_day1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="2023-02-16-blog-post-2_files/figure-html/descriptive-1.png" class="img-fluid figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Intake on second assessment (on a given day, non-consecutive)</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="do">## mean intakes</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(VF_simstudy<span class="sc">$</span>VF_day2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5.048887</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">## range (min to max value)</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(VF_simstudy<span class="sc">$</span>VF_day2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  0.00000 11.01431</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show relationship </span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>VF_simstudy,<span class="fu">aes</span>(<span class="at">x=</span>VF_day1,<span class="at">y=</span>LDL_cholesterol)) <span class="sc">+</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.3</span>,<span class="at">shape=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">"loess"</span>,<span class="at">se=</span><span class="cn">FALSE</span>,<span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span> </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Simulated relationship between vegetables and fruits and LDL cholesterol"</span>,</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle=</span><span class="st">"Vegetables and fruits intakes are measured on a given day"</span>,</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">caption=</span><span class="st">"n=5,000"</span>) <span class="sc">+</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="2023-02-16-blog-post-2_files/figure-html/descriptive-2.png" class="img-fluid figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="hand-calculation" class="level3">
<h3 class="anchored" data-anchor-id="hand-calculation">Hand calculation</h3>
<p>For the present demonstration of one independent variable measured with error (vegetables and fruits) in a linear regression model, the biased regression coefficient of the naive analysis can be calibrated directly using the following equation (Keogh and White, 2020):</p>
<p><span class="math display">\[ \beta_X = \beta\prime_{W_J}/\lambda\]</span> which reads as: the <em>unbiased</em> (calibrated) regression coefficient of the association between vegetables and fruits and LDL cholesterol <span class="math inline">\(\beta_X\)</span> equals to the <em>biased</em> (naive) regression coefficient obtained with measured vegetables and fruits on a given day <span class="math inline">\(\beta\prime_{W_J}\)</span> divided by the correction factor <span class="math inline">\(\lambda\)</span>.</p>
<p>The question is then, how can we estimate the correction factor <span class="math inline">\(\lambda\)</span>? This is the purpose of regression calibration. As shown in Keogh and White (2020), we can estimate the correction factor <span class="math inline">\(\lambda\)</span> by the regression of the replicate measurement <span class="math inline">\(W_{j=2}\)</span> on the first measurement <span class="math inline">\(W_{j=1}\)</span>. The equation is:</p>
<p><span class="math display">\[ W_{i,j=2} = \beta_0 + \boldsymbol{\lambda} W_{i,j=1} +\gamma Z_i+\epsilon_i\]</span> For the current demonstration, there are no covariates <span class="math inline">\(Z_i\)</span> and the equation simplifies to <span class="math inline">\(W_{i,j=2} = \beta_0 + \boldsymbol{\lambda} W_{i,j=1} +\epsilon_i\)</span>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co">#     Hand calculation of correction factor      #</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Regression of the second measurement (repeat) on first measurement</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>rc <span class="ot">&lt;-</span> </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glm</span>(<span class="at">formula =</span> VF_day2 <span class="sc">~</span> VF_day1, <span class="co"># had we considered covariates, we should have included them</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">family  =</span> <span class="fu">gaussian</span>(<span class="st">"identity"</span>),</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">data    =</span> VF_simstudy)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Show parameters</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  parameters<span class="sc">::</span><span class="fu">model_parameters</span>(rc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Parameter   | Coefficient |   SE |       95% CI | t(4998) |      p
------------------------------------------------------------------
(Intercept) |        3.09 | 0.07 | [2.96, 3.23] |   44.71 | &lt; .001
VF day1     |        0.39 | 0.01 | [0.37, 0.42] |   29.74 | &lt; .001</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Output correction factor (lambda)</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  lambda <span class="ot">&lt;-</span> parameters<span class="sc">::</span><span class="fu">model_parameters</span>(rc)[<span class="dv">2</span>,<span class="dv">2</span>]</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Show correction factor</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  lambda</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3920149</code></pre>
</div>
</div>
<p>Finally, we can calibrate the regression coefficient of the naive analysis (<span class="math inline">\(\beta\prime_{W_J=1}=\)</span> -0.04) by dividing it by the correction factor <span class="math inline">\(\lambda=\)</span> 0.392. Once calibrated, we find that the regression coefficient should be -0.102, consistent with the true association!</p>
</section>
<section id="the-mecor-package" class="level3">
<h3 class="anchored" data-anchor-id="the-mecor-package">The <code>mecor</code> package</h3>
<p>Fortunately, we don’t have to calculate everything by hand. The <code>mecor</code> R package (Nab et al.&nbsp;2021) can be used to calibrate a regression coefficient using repeated data of the independent variable measured with error. Using <code>mecor</code> is more efficient than the hand calculation above, as the package accommodates a wide range of scenarios and incorporates variance estimation.</p>
<p>The code below illustrates how to apply regression calibration with <code>mecor</code>. For demonstration purpose, the <code>method=standard</code> is used to show results consistent with traditional regression calibration, but other relevant methods can be used.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co">#          Reg. calibration with mecor           #</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ********************************************** #</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mecor)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>model_mecor <span class="ot">&lt;-</span> </span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>mecor<span class="sc">::</span><span class="fu">mecor</span>(<span class="at">formula =</span> LDL_cholesterol <span class="sc">~</span>    <span class="co"># outcome</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>               <span class="fu">MeasError</span>(VF_day1, <span class="at">replicate =</span> (VF_day2)) , <span class="co"># replicate measurement</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">method =</span> <span class="st">"standard"</span>,           <span class="co"># standard = regression calibration</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">data =</span> VF_simstudy)            <span class="co"># data</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>model_mecor</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
mecor::mecor(formula = LDL_cholesterol ~ MeasError(VF_day1, replicate = (VF_day2)), 
    data = VF_simstudy, method = "standard")

Coefficients Corrected Model:
(Intercept) cor_VF_day1 
  3.0064694  -0.1019799 

Coefficients Uncorrected Model:
(Intercept)     VF_day1 
 2.69087911 -0.03997762 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show confidence intervals for the corrected estimate (delta method)</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>(model_mecor)<span class="sc">$</span>c<span class="sc">$</span>ci</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             Estimate       LCI       UCI
(Intercept)  3.006469  2.890262  3.122677
cor_VF_day1 -0.101980 -0.124810 -0.079150</code></pre>
</div>
</div>
<p>Upon application of the measurement error correction (via standard regression calibration), we find that the coefficient of regression for the association between vegetables and fruits and LDL cholesterol is -0.102 mmol/L per 1-serving increase in vegetables and fruits intake. This is almost exactly the true association, as per the simulation, and equal to the ‘manual’ analysis above.</p>
<p><a href="https://didierbrassard.github.io/posts/2022/10/blog-post-4/">Bootstrap variance estimation</a> can also be performed within <code>mecor</code> with the <code>B=...</code> option.</p>
</section>
</section>
</section>
<section id="conclusion" class="level1">
<h1>Conclusion</h1>
<p>Thanks to Nab et al.&nbsp;(2021), regression calibration is relatively simple to implement in <code>R</code>, as well as other types of measurement error correction. I must point out that dietary intake data often have other issues that were not dealt with in this simple demonstration. For example, dietary intake data are often skewed which may require transformation to satisfy the classical measurement error assumptions. Covariates were also not considered, but they would have to be included in the regression calibration model (Keogh and White 2014, Keogh et al.&nbsp;2020). Also, study objectives are often much more complex; for example, many variables measured with error are often considered together in the same model. Nonetheless, I think the simple example above illustrates the importance of considering the impact of random errors and the relevance of measurement error correction methods.</p>
<p>For nutrition analyses in particular, I find the <a href="https://prevention.cancer.gov/research-groups/biometry/measurement-error-impact/software-measurement-error">NCI methods</a> currently provide the most comprehensive measurement error correction software. However, these methods are currently available in <code>SAS</code> only.</p>
</section>
<section id="reference" class="level1">
<h1>Reference</h1>
<p>Djousse, L., Arnett, D. K., Coon, H., Province, M. A., Moore, L. L. &amp; Ellison, R. C. (2004) Fruit and vegetable consumption and LDL cholesterol: the National Heart, Lung, and Blood Institute Family Heart Study. Am J Clin Nutr, 79(2), 213-7. doi:10.1093/ajcn/79.2.213.</p>
<p>Keogh, R. H. &amp; White, I. R. (2014) A toolkit for measurement error correction, with a focus on nutritional epidemiology. Stat Med, 33(12), 2137-55. doi:10.1002/sim.6095.</p>
<p>Keogh, R. H., Shaw, P. A., Gustafson, P., Carroll, R. J., Deffner, V., Dodd, K. W., Kuchenhoff, H., Tooze, J. A., Wallace, M. P., Kipnis, V. &amp; Freedman, L. S. (2020) STRATOS guidance document on measurement error and misclassification of variables in observational epidemiology: Part 1-Basic theory and simple methods of adjustment. Stat Med, 39(16), 2197-2231. doi:10.1002/sim.8532.</p>
<p>Kirkpatrick, S. I., Guenther, P. M., Subar, A. F., Krebs-Smith, S. M., Herrick, K. A., Freedman, L. S. &amp; Dodd, K. W. (2022) Using short-term dietary intake data to address research questions related to usual dietary intake among populations and subpopulations: Assumptions, statistical techniques, and considerations. J Acad Nutr Diet, 122(7), 1246-1262. doi:10.1016/j.jand.2022.03.010.</p>
<p>Nab, L., van Smeden, M., Keogh, R. H. &amp; Groenwold, R. H. H. (2021) Mecor: An R package for measurement error correction in linear regression models with a continuous outcome. Comput Methods Programs Biomed, 208, 106238. doi:10.1016/j.cmpb.2021.106238. PMID:34311414.</p>
<p>Thompson, F. E., Kirkpatrick, S. I., Subar, A. F., Reedy, J., Schap, T. E., Wilson, M. M. &amp; Krebs-Smith, S. M. (2015) The National Cancer Institute’s Dietary Assessment Primer: A Resource for Diet Research. J Acad Nutr Diet, 115(12), 1986-95. doi:10.1016/j.jand.2015.08.016.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>